### PR TITLE
Improve output from deployApp()

### LINKED ISSUE #669: More information from `deployApp()`
* If `appFiles` not specified, output which files are uploaded
* In `deploymentTarget()` output if using existing deployment (with `appId`) or if it's new. Might be worth outputting fully reified `account` + `server`. And need `appTitle` too. Maybe `createDeploymentTarget()` should report everything if `verbose`?

For QA, need to use that information to verify a few publishing workflows to make sure that we don't clash with how the IDE sets values. Need to check both new, republish, and "other destination" workflows.

### PR DESCRIPTION
**Pull request by hadley**
* Add some bigger structure with `cli::cat_rule()`
* Inform about appName and destination
* Don't display bundle/application ID so many times
* List uploaded files
* Strip timestamps from connect logs, highlight headings, and display on stdout, not stderr

Fixes #669

### CODE REVIEW

File: R/deployApp.R [hunk 1]
```diff
@@ -191,7 +191,7 @@ deployApp <- function(appDir = getwd(),
     check_directory(appSourceDoc)
     recordDir <- appSourceDoc
   } else if (!is.null(recordDir)) {
-    check_directory(recordDir)
+    check_file(recordDir)
```
Comment (hadley) [inline R/deployApp.R:194]: Badly named argument strikes again

File: R/deployApp.R [hunk 2]
```diff
@@ -288,23 +292,26 @@ deployApp <- function(appDir = getwd(),
     python <- getPythonForTarget(python, accountDetails)
     pythonConfig <- pythonConfigurator(python, forceGeneratePythonEnvironment)
 
+    if (!quiet) {
+      cli::cli_alert_info("Bundling files: {.file {appFiles}}")
```
Comment (hadley) [inline R/deployApp.R:296]: cli will automatically truncate this list if it's long

File: R/connect.R [hunk 3]
```diff
@@ -163,15 +163,17 @@ connectClient <- function(service, authInfo) {
       while (TRUE) {
         path <- paste0(file.path("/tasks", taskId), "?first_status=", start)
         response <- handleResponse(GET(service, authInfo, path))
+        messages <- unlist(response$status)
 
-        labeledMessage <- function(msg) {
-          message(paste("[Connect]", msg))
-        }
+        # Strip timestamps, if found
+        timestamp_re <- "\\d{4}/\\d{2}/\\d{2} \\d{2}:\\d{2}:\\d{2}\\.\\d{3,} "
```
Comment (aronatkins) [inline R/connect.R:169]: The `[Connect]` preamble was trying to make it more clear what information was produced on the client and what information was sent back from the server. Do you feel like this separation is still clear?

File: R/deployApp.R [hunk 4]
```diff
@@ -288,23 +292,26 @@ deployApp <- function(appDir = getwd(),
...
-        contentCategory = contentCategory,
-        verbose = verbose,
-        pythonConfig = pythonConfig,
-        quarto = quarto,
-        isCloudServer = isCloudServer(accountDetails$server),
-        metadata = metadata,
-        image = image
-      )
-
+    withStatus("Uploading bundle", {
```
Comment (aronatkins) [inline R/deployApp.R:314]: Is it silly to have `logger` immediately followed by `withStatus`, which also logs?

File: R/deployApp.R (hunk 2 above)
Comment (aronatkins) [reply R/deployApp.R:296]: `bundleApp` can be (relatively) expensive. It would be nice to have similar to `withStatus` which marks the start/end of this type of operation (with timestamps). The existing `withStatus` is a little too awkward for this purpose, especially since `bundleApp` can produce output -- its "DONE" message expects it is emitted on the same line as the original status message.

Not for this PR, for sure; idle musing about better ways to mark the start and end of different processing phases.

File: R/deployApp.R [hunk 5]
```diff
@@ -324,29 +331,28 @@ deployApp <- function(appDir = getwd(),
     bundle <- application$deployment$bundle
   }
 
-  if (length(bundle$id) > 0 && nzchar(bundle$id)) {
```
Comment (aronatkins) [inline R/deployApp.R:327]: Do you think that we still log enough (early) information about the deployment target? Output from rsconnect is often attached to support tickets, and breadcrumbs like these are occasionally helpful.

**Review by aronatkins — APPROVED** [review]

File: R/connect.R (hunk 3 above)
Comment (hadley) [reply R/connect.R:169]: I think so, because the rules help delineate the output from connect:

Screenshot 2023-02-28 at 09 01 21

Maybe the rule could have a more informative title? I can't think what it should say though.

File: R/deployApp.R (hunk 2 above)
Comment (hadley) [reply R/deployApp.R:296]: Oh I assumed that `bundleApp()` would be cheap. Is it just the cost of tar-ing up a bunch of files?

In general, would you be ok with moving to something that looks more like:

```
* Bundling files...
✅ Bundling complete
```

The dangling line makes all sorts of output (including errors) look wonky.

File: R/deployApp.R (hunk 5 above)
Comment (hadley) [reply R/deployApp.R:327]: Yeah, let me try again.

File: R/deployApp.R (hunk 2 above)
Comment (aronatkins) [reply R/deployApp.R:296]: `bundleApp` includes manifest creation, which means dependency discovery (finding explicit package references, then exploding the entire package requirement tree). It has been expensive in the past. https://github.com/rstudio/packrat/pull/615, for example.

**Comment by hadley** [comment]
I'm sure there's still more work to be done, but I think this looks pretty good and it sets up a good foundation for the future:

Screenshot 2023-02-28 at 11 19 50
